lib/Controller/PasswordController.php

```diff
@@ -0,0 +1,177 @@
+use OCP\IL10N;
+use OCP\IRequest;
+use OCP\ISession;
+use OCP\IURLGenerator;
+use OCP\IUserManager;
+use OCP\IUserSession;
+
+class PasswordController extends Controller implements IAccountModuleController {
+
+	/** @var IUserSession */
+	protected $userSession;
+	/** @var IUserManager */
+	protected $userManager;
+	/** @var IConfig */
+	protected $config;
+	/** @var ISession */
+	protected $session;
+	/** @var IURLGenerator */
+	protected $urlGenerator;
+	/** @var IL10N */
+	protected $l10n;
+
+	/**
+	 * PasswordController constructor.
+	 *
+	 * @param $appName
+	 * @param IRequest $request
+	 * @param IUserSession $userSession
+	 * @param IUserManager $userManager
+	 * @param IConfig $config
+	 * @param ISession $session
+	 * @param IURLGenerator $urlGenerator
+	 * @param IL10N $l10n
+	 */
+	public function __construct(
+		$appName,
+		IRequest $request,
+		IUserSession $userSession,
+		IUserManager $userManager,
+		IConfig $config,
+		ISession $session,
+		IURLGenerator $urlGenerator,
+		IL10N $l10n
+	) {
+		parent::__construct($appName, $request);
+		$this->userSession = $userSession;
+		$this->userManager = $userManager;
+		$this->config = $config;
+		$this->session = $session;
+		$this->urlGenerator = $urlGenerator;
+		$this->l10n = $l10n;
+	}
+
+	/**
+	 * @param $redirect_url
+	 * @param string $error (optional)
+	 * @return TemplateResponse
+	 */
+	protected function createPasswordTemplateResponse($redirect_url, $error = null) {
+		$params = [
+			'redirect_url' => $redirect_url,
+			'webroot' => \OC::$WEBROOT
+		];
+		if (empty($params['redirect_url'])) {
+			$params['redirect_url'] = $this->urlGenerator->getAbsoluteURL('');
+		}
+		if ($error) {
+			$params['error'] = $error;
+		}
+		return new TemplateResponse(
+			$this->appName,
+			'password',
+			$params,
+			'guest'
+		);
+	}
+	/**
+	 * Shows a change password template
+	 *
+	 * @NoCSRFRequired
+	 *
+	 * @param string $redirect_url
+	 * @return TemplateResponse|JSONResponse
+	 *
+	 * @throws \Exception
+	 */
+	public function show($redirect_url) {
+		// gracefully handle redirected ocs api requests
+		if ($this->request->getHeader('OCS-APIREQUEST') === 'true') {
+			return new JSONResponse([], Http::STATUS_NO_CONTENT);
+		}
+		return $this->createPasswordTemplateResponse($redirect_url);
+	}
+
+	/**
+	 * Shows a change password template
+	 *
+	 * @param string $current_password
+	 * @param string $new_password
+	 * @param string $retyped_password
+	 * @param string $redirect_url
+	 * @return TemplateResponse|RedirectResponse
+	 *
+	 * @UseSession
+	 * @throws \Exception
+	 */
+	public function update($current_password, $new_password, $retyped_password, $redirect_url) {
+		if ($new_password !== $retyped_password) {
+			return $this->createPasswordTemplateResponse(
+				$redirect_url,
+				$this->l10n->t('New passwords are not the same.')
+			);
+		}
+
+		$user = $this->userSession->getUser();
+		if ($user === null) {
+			throw new \UnexpectedValueException('No user in session');
```

Comment:
How about: "_The user is not currently logged in._".

Reply:
This should never happen because the method is not annotated with `@PublicPage` so a user must be logged in. In turn the session must have a user set. I'll just delete this check because it only makes the code inspection happy, anyway.